WordPress/WordPressTest/CoreDataStackHelper.swift

```diff
@@ -0,0 +1,49 @@
+import Combine
+import XCTest
+
+@testable import WordPress
+
+extension ContextManagerMock {
+    /// Override `ContextManager.shared` with the mock instance, and un-override it when
+    /// given `testCase` finishes.
+    ///
+    /// This method should be called during test run, typically from `XCTestCase` subclass's
+    /// `setUp` method or test methods.
+    ///
+    /// - Parameter testCase: The test case to wait for.
+    @objc func useAsSharedInstanceUntilTestFinished(_ testCase: XCTestCase) {
+        setUpAsSharedInstance()
+
+        var cancellable: AnyCancellable?
+        cancellable = AutomaticTeardownTestObserver.instance.testCaseDidFinishPublisher
```

Comment:
Tip: using `Combine` + all the dance with your custom `AutomaticTeardownTestObserver` + unusual cancellable pattern… can all be avoided and greatly simplified by just using [this method of XCTestCase instead](https://developer.apple.com/documentation/xctest/xctestcase/3815521-addteardownblock:) 🙂 

```swift
    @objc func useAsSharedInstance(untilTestFinished testCase: XCTestCase) {
         setUpAsSharedInstance()
         testCase.addTeardownBlock { [weak self] self?.tearDown() }
     }
```

Reply:
Btw, once we have that working, I'm not sure if it's really worth (nor a good idea, at the risk of API misuse) to keep `self.setUpAsSharedInstance()` and `self.tearDown` as public API of `ContextManagerMock` after all.

I mean, I think we could instead just have this as public API:
```objc
// .h
- (void)useAsSharedInstanceUntilTestFinished:(XCTestCase*)testCase NS_SWIFT_NAME(useAsSharedInstance(untilTestFinished:));

// .m
- (void)useAsSharedInstanceUntilTestFinished:(XCTestCase*)testCase
{
    ContextManager* original = [ContextManager internalSharedInstance];
    [ContextManager overrideSharedInstance:self];
    [testCase addTeardownBlock:^{
        [self.mainContext reset];

        if ([ContextManager sharedInstance] == self) {
            [ContextManager overrideSharedInstance:original];
        }
    }];
 }
```
Then get rid of `-(void)setUpAsSharedInstance` and `-(void)tearDown` to avoid the risk of people calling the latter without calling the latter to balance it.

Given this solution using `addTeardownBlock` does no longer have the limitations of the "Closure-based API" that you mention in "Other API suggestions" (which might not be as easily used if the start and end of the test code were not as "linear" as with a synchronous imperative code, and which was the initial reason we thought about keeping the manual separate `setUpAsSharedInstance`/`tearDown` dance available when needed)… that means that now just exposing `useAsSharedInstance(untilTestFinished:)` should be enough to cover all cases (sync / imperative / linear test code, and async tests) AND ensure that cleanup is done properly in all cases without risk of API misuse or test leakage.

Reply:
Oh nice, I didn't know this `addTeardownBlock` API exists. That'll make things much simpler!

Reply:
I've reimplemented this method using `addTeardownBlock`, it works perfectly, almost, except one issue.

From [the Apple doc](https://developer.apple.com/documentation/xctest/xctestcase/set_up_and_tear_down_state_in_your_tests#3909111):
> If you register any teardown blocks during a test method’s execution, XCTest runs them after that test method ends but before XCTest calls the teardown instance method.

Which means `ContextManager.shared` is the mock instance _before_ `XCTestCase.tearDown()`, but the real singleton _during_ the execution of `XCTestCase.tearDown`. I guess we can express this behaviour by renaming the method name from `useAsSharedInstance(untilTestFinished:)` to `useAsSharedInstance(beforeTestTorndown:)`, but we still need to be mindful that `ContextManager.share` references two different object during test execution.

Considering `XCTestCase.tearDown` is the place for clean up work, also the fact that a brand new in memory database is created for each test run, it might not be necessary to do any "clean up" to a soon to be thrown away in-memory database using `ContextManager.shared` inside `XCTestCase.tearDown`. I think it's fine to expect `ContextManager` not being used at all inside `XCTestCase.tearDown`, but it might be hard to enforce this expectation via API.

But still `ContextManager.shared` might be implicitly called during `XCTestCase.tearDown`. Like [this line here](https://github.com/wordpress-mobile/WordPress-iOS/blob/c16a40a3d23f186a28f5176b69b22c6ea395234a/WordPress/WordPressTest/Dashboard/DashboardCardTests.swift#L28), which uses `ContextManager.shared` underneath to make changes to the `blog` object. For the reason I discussed above, it's probably okay to delete this line, unless there is also some clean up work needed to be done to the `QuickStartTourGuide.shared` singleton itself. Either way, this is one example that there are still chances `ContextManager.shared` being used implicitly during `XCTestCase.tearDown`.

BTW, the particular line referred above actually can't use the real singleton, due to a deadlock issue, which is "fixed" in [the mock](https://github.com/wordpress-mobile/WordPress-iOS/blob/c16a40a3d23f186a28f5176b69b22c6ea395234a/WordPress/WordPressTest/ContextManagerMock.m#L70-L72). I had discussed this issue with @mokagio, and we decided to park the investigation on this issue for my trial project.

IMHO, it would be ideal to make this function work through the whole execution of a test, rather than before `tearDown`. But I also don't see any huge issue with limiting it before `tearDown`, it's probably best to express this behaviour explicitly to avoid any surprises, like renaming the function. If we go with the later option, we'll need to address the deadlock issue, which could be simply deleting that line. What do you think?

Reply:
Good point about the code in custom `teadDown()` methods being able to also call the `sharedInstance` after the mock has been removed and creating a real instance for nothing.

Another alternative would be to make the `addTeardownBlock` code not restore the original and true `sharedInstance` but instead make the current `sharedInstance` invalid to be called, so that if it gets ultimately called it would e.g. crash with a message telling the developers that this is invalid (and that if they really really need to do something with the mock context before it's removed — however unlikely given it's an in-memory instance anyway, but 🤷  — they should instead add their own `addTeardownBlock` after installing the mock (the teardown blocks being called in reverse order of their declaration at the end of a test)

I'm not sure what would be the best way to implement such a trick though.

The easiest way could be to implement some flag like `BOOL markAsInvalidToUse` or something and then make the `sharedInstance` method (that would otherwise return the overridden `sharedInstance` / mock) crash with some `fatalError` with an explicit message if that flag was ON. Then turn the flag on in the `addTeardownBlock` call so that any subsequent call would crash. But I think that would require to implement the flag on the `ContextManager` itself, rather than the `ContextManagerMock`, which means introducing test-specific code on the real-world object, which is not ideal. Sure, we already do this with `ContextManager.m` having the `_override` static while this is only used for test code, so we already use that twisted pattern, but that doesn't mean we should continue in that territory 😓 

Another way would be to create a proxy object whose `repondToSelector:` or similar runtime method would be implemented to return such error message on any method call. We'd then make `addTeardownBlock` switch the `ContextManagerMock` with this proxy `InvalidContextManager` instance so that any subsequent use would crash with the message telling devs about their invalid use.

I'm not sure either solution is the good one nor if it's really worth the trouble, but I figured I'd still throw those ideas in.

Reply:
My instinct(only based on my experience with the project in past couple weeks, I could very much be wrong) about this approach is there are probably a few test cases will actually hit this crash. The test case itself probably doesn't actively call `ContextManager.sharedInstance` to do something, but the "test-ee" may have some async operations which use the `ContextManager` singleton, and it's very likely these async operations are going to be executed after the test method. For this scenario where the application code continues doing its thing after test completes, I would argue it's a reasonable thing to do and we probably shouldn't crash the test suite for it. It probably isn't easy for the application to stop using `ContextManager` singleton after a certain test finishes, nor for the test to wait till the application code done using the singleton to finish. Maybe some sort of warning would be nicer? Anyway, I can do a quick code change to see if such test case exists.

I just had an idea seeing you mentioned `respondToSelector:`. If the Combine dance looks a bit strange, I think we can replace it with a slightly less strange approach. We could tear down `ContextManagerMock` directly from the test observer. Something like this:
```swift
private class AutomaticTeardownTestObserver: NSObject, XCTestObservation {
    func testCaseDidFinish(_ testCase: XCTestCase) {
        testCase.additionalTeardown?()
    }
}

private extension XCTestCase {
    var additionalTeardown: (() -> Void)? {
        set {
            objc_setAssociatedObject(self, key, newValue, .retain)
        }
        get {
            objc_getAssociatedObject(self, key)
        }
    }
}

extension ContextManagerMock {
    @objc func useAsSharedInstanceUntilTestFinished(_ testCase: XCTestCase) {
        setUpAsSharedInstance()
        testCase.additionalTeardown = { [weak self] in
            self?.tearDown()
        }
    }
}
```

Reply:
That solution seems like an acceptable compromise to me.

I mean, tbh it still feels a bit workaround-y having to use a `TestObserver` and `objc_associatedObject` (regardless of me loving playing with ObjC runtime like that 😛 ) here as opposed to something built-in and with nicer official API like `addTeardownBlock`… but on the other end it would also solve the issue about `addTeardownBlock` being run before the actual `func tearDown()`… As long as we'd add a clear comment in the code explaining the rationale of why we're doing this dance as opposed to using `addTeardownBlock()` (to avoid a future us being tempted to replace it with `addTeardownBlock` thinking it would be simpler, without realizing the impact).